File: llm_generated_code/ChatGPT/Python/Level1/Q7.py
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

def has_cycle(head):
    if not head or not head.next:
        return False

    slow = head  # Moves one step at a time
    fast = head.next  # Moves two steps at a time

    while slow != fast:
        if not fast or not fast.next:
            return False  # Fast pointer has reached the end of the list
        slow = slow.next  # Move slow pointer one step
        fast = fast.next.next  # Move fast pointer two steps

    return True  # Slow and fast pointers have met, cycle exists

File: llm_generated_code/ChatGPT/CPP/Level1/Q7.cpp
#include <iostream>
using namespace std;

struct ListNode {
    int val;
    ListNode* next;
    ListNode(int x) : val(x), next(nullptr) {}
};

bool hasCycle(ListNode* head) {
    if (!head || !head->next) return false;

    ListNode* slow = head;
    ListNode* fast = head->next;

    while (fast != nullptr && fast->next != nullptr) {
        if (slow == fast) return true;
        slow = slow->next;
        fast = fast->next->next;
    }

    return false;
}

int main() {
    // Creating a linked list with a cycle for testing
    ListNode* head = new ListNode(1);
    head->next = new ListNode(2);
    head->next->next = new ListNode(3);
    head->next->next->next = new ListNode(4);
    head->next->next->next->next = head->next; // Creating a cycle

    if (hasCycle(head)) {
        cout << "Cycle detected in the linked list." << endl;
    } else {
        cout << "No cycle in the linked list." << endl;
    }

    return 0;
}

